Make solana-validator check vote account at start

ryoqun opened the pull request:
#### Problem

From #6628
> It's easy to forget to run `solana create-vote-account` before starting a validator and then get confused as to why it's not possible to `solana delegate-stake` to your vote account that doesn't exist. 


#### Summary of Changes

Again from #6628:
> We can help the user by having `solana-validator` ensure the vote account it was provided actually exists.

### Memo

~This should be backported to v0.20.3.~

Fixes #6628

ryoqun commented on core/src/validator.rs:
```diff
@@ -179,6 +180,9 @@ impl Validator {
         let bank_info = &bank_forks_info[0];
         let bank = bank_forks[bank_info.bank_slot].clone();
+
+        Self::check_vote_account(&bank, vote_account);
```
Created a method, this constructor is too long already...

ryoqun commented on validator/Cargo.toml:
```diff
@@ -17,6 +17,7 @@ log = "0.4.8"
 indicatif = "0.13.0"
 reqwest = { version = "0.9.22", default-features = false }
 serde_json = "1.0.41"
+solana-cli = { path = "../cli", version = "0.21.0" }
```
I'm not sure depending on "cli" from "validator" can be justified...

mvines replied:
I’d rather not have validator depend on cli.    It’s long overdue that we create a common clap-validators/ crate where we can easily share these validators between all the tools. 

 Care to add that common crate first and then do the input validator cleanup here?  We can do that in a separate PR from the check_vote_account() changes here.

ryoqun replied:
Heh, that's what I wanted to do ideally. I'll do the cleaning job now that I got the approval. :)

ryoqun replied:
Done at https://github.com/solana-labs/solana/pull/6812

ryoqun commented on validator/src/main.rs:
```diff
@@ -223,13 +227,6 @@ fn initialize_ledger_path(
 }
 
-// Return an error if a keypair file cannot be parsed.
-fn is_keypair(string: String) -> Result<(), String> {
```
This is small win resulting from [this](https://github.com/solana-labs/solana/pull/6790/files#r343502742).

ryoqun commented on core/src/validator.rs:
```diff
@@ -406,6 +410,20 @@ impl Validator {
+            } else {
+                error!("Vote account not found: {}:", vote_account);
+            }
+            process::exit(1);
```
I'm not very sure why validator aborting with `process::exit(1)` around this code. Maybe `panic_hook` related?

mvines replied:
I’m not sure I understand what you mean?

ryoqun replied:
Well, I mean, why is there several places calling `process::exit(1)` in `validator.rs`?:

```shell
ryoqun@ubuqun:~/work/solana/solana$ git grep -B4 -A1 process::exit core/
core/src/validator.rs-            if is_x86_feature_detected!("avx") {
core/src/validator.rs-                info!("AVX detected");
core/src/validator.rs-            } else {
core/src/validator.rs-                error!("Your machine does not have AVX support, please rebuild from source on your machine");
core/src/validator.rs:                process::exit(1);
core/src/validator.rs-            }
--
core/src/validator.rs-            error!(
core/src/validator.rs-                "Delete the ledger directory to continue: {:?}",
core/src/validator.rs-                blocktree_path
core/src/validator.rs-            );
core/src/validator.rs:            process::exit(1);
core/src/validator.rs-        }
--
core/src/validator.rs-        process_options,
core/src/validator.rs-    )
core/src/validator.rs-    .unwrap_or_else(|err| {
core/src/validator.rs-        error!("Failed to load ledger: {:?}", err);
core/src/validator.rs:        process::exit(1);
core/src/validator.rs-    });
```

I think [`panic!` is preferred to than `process::exit(1)` in Rust](https://stackoverflow.com/a/39229209/10242004):

> rather use panic!() in nearly all cases

Better yet, I think because `core` rather functions as a library code, so it's not graceful to use `process::exit` in the middle of code (or `panic!` in that matter). Preferably, we should return `Result<Validator>` from `Validator::new`? Of course, some amount of rewrite will be needed... 

At least, we should comment why `process.exit` is used at the very least. The reason of `process::exit` isn't clear to me...

mvines replied:
> Preferably, we should return Result<Validator> from Validator::new? Of course, some amount of rewrite will be needed...

Yes exactly.  `new` functions typically never fail, so a better way I think would be to replace `Validator::new()` with a `Validator::create()` method that returns a `Result`.  

Because we don't have that nice `Validator::create()` we're using `exit(1)` in those places where really the error ought to be bubbled up to the caller (some action the user could take to resolve the issue), and reserving `panic!()` for cases where we have an internal error.

This is another clean-up item that you're very welcome to take on if you'd like!

ryoqun replied:
Thanks for clarification! Then, `process::exit` is used just by historical reasons and I'm glad that cleaning up would be welcomed. :)

mvines replied:
Yep, you'll see that a lot around here.   A bad code pattern gets replicated around for a while until somebody finally says "Enough!  We need to clean up this mess" and just does it :)

ryoqun commented on validator/src/main.rs:
```diff
@@ -266,7 +263,7 @@ pub fn main() {
                 .value_name("PUBKEY")
                 .takes_value(true)
-                .validator(is_keypair)
+                .validator(is_pubkey_or_keypair)
```
Contrary to [this help message](https://github.com/solana-labs/solana/pull/6790/files#diff-1f422d7e73618aca7bce9935356de012R264), `--vote-account` never accepted a pubkey it seems...

ryoqun replied:
This is fixed by https://github.com/solana-labs/solana/pull/6812

mvines commented on core/src/validator.rs:
```diff
@@ -406,6 +410,20 @@ impl Validator {
+                    found_account.owner, vote_account
+                );
+            } else {
+                error!("Vote account not found: {}:", vote_account);
```
nit: remove the second : (after the {})

ryoqun commented:
Uuum, I hideously failed on CIs. Working...

ryoqun commented:
To work on [this](https://github.com/solana-labs/solana/pull/6790#discussion_r343505652), I'll submit another preparatory PR shortly after.

Remaining for this PR:
- Clarify comments
- Stop voting when there is no vote account at ReplayStage and reject self-vote with no vote account at Tower?

ryoqun commented on validator/src/main.rs:
```diff
@@ -562,13 +575,6 @@ pub fn main() {
         }
     };
 
-    solana_logger::setup_with_filter(
```
;) No wonder I couldn't see `debug!`s, `warn!`s, etc in the preceding large chunk of arg handling code....

mvines replied:
Yeah, I recently moved it here from where you moved it back to because I wanted the arg parsing code to use println/eprintln only.  The log crate then only gets used after the `--log` argument is processed and the log destination is configured.

mvines replied:
The entire argument parsing in this function eventually needs to be split out into it's own function, so the clippy complexity suppression can be removed

ryoqun replied:
Thanks for detailed explanation. I'll then just resotre this change.

mvines commented on core/src/validator.rs:
```diff
@@ -179,6 +180,13 @@ impl Validator {
+
+        // Only do this check if started as a bootstrap leader.
+        // Normal validators can encounter newly-created vote accounts, which to be synced with this new validator.
+        if entrypoint_info_option.is_none() && !config.voting_disabled {
```
Normal validators where the ones getting tripped up with the lack for vote account, not the bootstrap leader or a blockstreamer node.   How about we the vote account check moves from `core/` into `validator/`, and the check happens over RPC to the entrypoint (so the check does not occur for a bootstrap leader, which should be fine)

ryoqun replied:
Hm, that sounds nice! I was shorted sighted...

ryoqun replied:
@mvines I've tried to implement what you proposed [like this](https://github.com/solana-labs/solana/pull/6790/files?utf8=%E2%9C%93&diff=split&w=1#diff-1f422d7e73618aca7bce9935356de012R208). Could you just take a light look at it? If this approach looks good to proceed, I'll polish this up further including rebaseing on now-approved preparatory PR #6812 (thx for approvals, btw!).

Also, I rather want to keep the vote account check for bootstrapping nodes in the `./core/`. As this may sound redundant/excessive, I think this is useful to new prospective contributors fiddling the `solana-validator` command, and I like the completeness of the check for both cases of bootstrapping/normal nodes

mvines replied:
I think I'd prefer to remove the check from here to avoid the semi-duplication.  The bootstrap node is a little special, it's only started by one person and they're really sure about it, so I don't mind the lack of vote check for the bootstrap node

ryoqun replied:
I see. Just restored `./core/src/validator.rs` to the original state. :smile: 

Now this PR touches only `./validator/src/main.rs`.
/me feels safe :tongue:

ryoqun commented on validator/src/main.rs:
```diff
@@ -190,17 +193,42 @@ fn initialize_ledger_path(
+        Ok((rpc_addr, RpcClient::new_socket(rpc_addr)))
+    } else {
+        Err(format!(
+            "No node including entrypoint ({:?}) is running the RPC service",
```
A bit nicer message. We iterate over nodes in fact.

mvines replied:
I'm pretty sure we only talk to the entrypoint's RPC port here.  Where do you see otherwise?

ryoqun replied:
Yeah, I was wrong. I got misconception by `filter_map`. So, I changed to `find_map`: https://github.com/solana-labs/solana/pull/6790/files#diff-1f422d7e73618aca7bce9935356de012R185

According to the rust doc, these should be equivalent: https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.find_map

ryoqun commented on validator/src/main.rs:
```diff
@@ -221,23 +249,24 @@ fn initialize_ledger_path(
-
 #[allow(clippy::cognitive_complexity)]
 pub fn main() {
+    solana_logger::setup_with_filter(
```
TODO to @ryoqun: move to the original place.

ryoqun replied:
done

mvines commented on validator/src/main.rs:
```diff
@@ -190,17 +193,42 @@ fn initialize_ledger_path(
+        .map_err(|err| err.to_string())?;
+
+    if found_vote_account.owner == solana_vote_api::id() {
+        Ok(())
```
Let's add additional checks in here for these two vote account fields:
https://github.com/solana-labs/solana/blob/f54cfcdb8f2c6f44f80c1c7b6ec02fb2e71fa3e3/programs/vote_api/src/vote_state.rs#L87-L89

ryoqun replied:
Done. Maybe is this something you wanted to check? https://github.com/solana-labs/solana/pull/6790/files#diff-1f422d7e73618aca7bce9935356de012R223

mvines commented:
@ryoqun - hey I don't think we need this in 0.20 after all, 0.21 will be coming soon enough :)

ryoqun commented:
> @ryoqun - hey I don't think we need this in 0.20 after all, 0.21 will be coming soon enough :)

I see!

ryoqun commented:
@mvines I've polished this up to the finished work! I think this is getting pretty close to the merge-ready. :smile:

ryoqun commented on validator/src/main.rs:
```diff
@@ -183,28 +182,73 @@ fn initialize_ledger_path(
+    voting_pubkey: &Pubkey,
+    node_pubkey: &Pubkey,
+) -> Result<(), String> {
+    let found_account = rpc_client
```
@mvines Come to think of this richness of sanity checks, I think checking minimum lamports in `node_pubkey` might also be helpful. Should add that? Is there any concern?

mvines replied:
Yes that'd be great.  I don't know what a minimum should be right now so let's just check for non-zero as a start

ryoqun replied:
I ended up checking existince of node account and account balance being at least `1 lamport`. After all there should be no account with `0 lamports`. Although, I had [this bug](#6890) in my mind.

mvines commented on validator/src/main.rs:
```diff
@@ -641,23 +697,40 @@ pub fn main() {
             }
+            validator_config.expected_genesis_hash = Some(genesis_hash);
+        } else {
+            error!("unable to create rpc client");
```
Can your print the error itself along side the `unable to create rpc client` message?  It looks like we just consume the error and don't show the user?

ryoqun replied:
Nice catch! https://github.com/solana-labs/solana/pull/6790/files#diff-1f422d7e73618aca7bce9935356de012R725

ryoqun commented:
@mvines Rebased, and addressed your last 2 comments, and fixed [an embarrassing message bug](https://github.com/solana-labs/solana/pull/6790/commits/8e5e8dafe2c89decd97b9c3861f3e626c23edbd2). :scream: 

To make mergify happy, I requested another review request!

CI is green; Can I land this? :smile:
